extensions/graphql/src/graphql.container.ts

```diff
@@ -0,0 +1,90 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/graphql
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  Binding,
+  Constructor,
+  Context,
+  createBindingFromClass,
+  filterByKey,
+  filterByServiceInterface,
+} from '@loopback/core';
+import debugFactory from 'debug';
+import {ContainerType, ResolverData} from 'type-graphql';
+import {GraphQLBindings, GraphQLTags} from './keys';
+
+const debug = debugFactory('loopback:graphql:container');
+
+/**
+ * Context for graphql resolver resolution
+ */
+export class GraphQLResolutionContext extends Context {
+  constructor(
+    parent: Context,
+    readonly resolverClass: Constructor<unknown>,
+    readonly resolverData: ResolverData<unknown>,
+  ) {
+    super(parent);
+    this.bind(GraphQLBindings.RESOLVER_DATA).to(resolverData);
+    this.bind(GraphQLBindings.RESOLVER_CLASS).to(resolverClass);
+  }
+}
+
+/**
+ * Implementation of `ContainerType` to plug into `type-graphql` as the IoC
+ * container
+ */
+export class LoopBackContainer implements ContainerType {
+  constructor(readonly ctx: Context) {}
+  get(
+    resolverClass: Constructor<unknown>,
+    resolverData: ResolverData<unknown>,
+  ) {
+    debug('Resolving a resolver %s', resolverClass.name, resolverData);
+    const resolutionCtx = new GraphQLResolutionContext(
```

Comment:
IIUC, you are creating a new resolution context for each resolver class we are instantiating. Personally, I would expect to have a context per incoming request, either instead or in addition to per-resolver-class context.

For example, when executing the example query shown earlier in your pull request, how many context instances are created?

```graphql
query {
  recipe(recipeId: "1") {
     title
     description
     ratings
     creationDate
     ratingsCount(minRate: 2)
     averageRating
     ingredients
     numberInCollection
   }
 }
}
```

More importantly, how can the application developer share state between different resolver instances created while handling a single incoming request? In REST API, it's possible to share state via per-request context (e.g. the authentication layer sets the current user, the authorization layer reads that value),

Reply:
The `GraphQLResolutionContext` is only created to contain the resolver binding if it does not exist in the chain. At the moment, we're building a GraphQL schema (and LoopBack container for DI) per application.